apps/SoapySDRUtil.1
.\" SoapySDRUtil.1 - manpage for SoapySDRUtil
.\"
.\"
.TH SOAPYSDRUTIL 1 2016\-10\-28 "SoapySDR 0.5.4"
.SH NAME
\fBSoapySDRUtil\fR \- query and change module information of SoapySDR
.\" ----------------------------------------------------------------------------
.SH SYNOPSIS
\fBSoapySDRUtil\fR [\fIOPTIONS\fR]
.\" ----------------------------------------------------------------------------
.SH DESCRIPTION
SoapySDRUtil is a command line application that provides basic
diagnostic utility for SoapySDR by printing installation information,
available modules, discovered devices, and tests on instantiated devices.
.\" ----------------------------------------------------------------------------
.SH OPTIONS
.TP
\fB\-\-info\fR
Print general information on the library, list all found hardware support
modules and available factories.
.TP
\fB\-\-find\fR[="\fISPEC\fR"]
Discover available devices, restricted to those matching the \fISPEC\fR if
given.
If no matching devices are found it will exit with exit status 1.
.TP
\fB\-\-make\fR[="\fISPEC\fR"]
Open and close device matching \fISPEC\fR if possible.
If \fISPEC\fR is not given, the first discovered device is opened.
.TP
\fB\-\-probe\fR[="\fISPEC\fR"]
Show detailed information on all devices or the devices matching \fISPEC\fR if
given.
.TP
\fB\-\-check\fR=\fINAME\fR
Check and print if driver module named \fINAME\fR is present.
If it is not found it will exit with exit status 1.
.\" ----------------------------------------------------------------------------
.SH HOMEPAGE
SoapySDRUtil is part of the
.UR https://github.com/pothosware/SoapySDR/wiki
SoapySDR project
.UE .
.\" ----------------------------------------------------------------------------
.SH AUTHORS
The SoapySDR library, including SoapySDRUtil, was written mostly by
.MT josh@pothosware.com
Josh Blum
.ME .
.\"
.\"This man page was originally written for the Debian project by
.\".MT aeb@debian.org
.\"Andreas Bombe
.\".ME .


Changelog.txt
This this the changelog file for the SoapySDR project.

Release 0.6.0 (pending)
==========================

Device C API changes:

- Device readI2C - make numBytes an in/out argument
- Device read/write registers switch to named variant
- Status return for Device C API calls with void return

General additions and changes:

- Added rate testing to SoapySDRUtil application
- Added listSearchPaths() API and SoapySDRUtil print
- Added Kwargs type to/from markup string API calls
- Added read/writeRegisters() API for bulk register IO

Release 0.5.4 (pending)
==========================

- Added man page for SoapySDRUtil executable
- Dynamic root environment variable to support snappy packages
- Windows: disable graphical error pop-ups when DLLs are not found
- Debian: replaced recommends with soapysdr-modules-all metapackage

Release 0.5.3 (2016-09-01)
==========================

- Additional try/catch blocks for C API wrapper safety

Release 0.5.2 (2016-08-18)
==========================

- Fixes for Doxygen warnings and Doxyfile version upgrade
- Additional build helpers for python development config

Release 0.5.1 (2016-08-02)
==========================

- Added SoapySDR_getLibVersion() and SOAPY_SDR_API_VERSION
- debian allow any swig >= 2.0.0 using the swig meta-package
- Only use stderr in the default logger for command line tools
- ABI-specific modules directory for multi-version install

Release 0.5.0 (2016-07-10)
==========================

Deprecated Device API calls:

- Deprecated setCommandTime, use setHardwareTime
- Deprecated listBandwidths, use getBandwidthRange
- Deprecated unnamed read/writeRegister, use named

General additions and changes:

- Created wait trigger flag for trigger-based streaming
- Added device factory lock to the python enumerate wrapper
- Added named register API for multiple register interfaces
- Added channel info API call for channel-specific info
- Setup CMake rpath handling to be enabled by default
- Moved time source calls to time API section
- C bindings return status code and stash the error message
- Added getBandwidthRange() API call for continuous ranges
- Use current DLL path to determine the system install root
- Implemented channel-specific settings API

Release 0.4.4 (2016-07-10)
==========================

- Add known python version 3.5 to FindPython3Libs.cmake
- Fix FindPython3Libs.cmake use of python3-config --includes

Release 0.4.3 (2016-04-21)
==========================

- CMake SOAPY_SDR_SOVER variable cache fix use INTERNAL type
- Added missing memset for internal use of SoapySDRKwargs type
- Fix SoapySDRDevice declarations for use in a C only compiler

Release 0.4.2 (2016-03-02)
==========================

- Fixes for SOAPY_SDR_MODULE_UTIL destination and install prefix
- CMake override for PYTHON_INSTALL_DIR and PYTHON3_INSTALL_DIR

Release 0.4.1 (2016-01-27)
==========================

- Added readStreamStatus() to python bindings
- Notes on stream timeouts for client code compatibility
- Fix debian upstream changelog and license install
- Added official modules to debian Recommends

Release 0.4.0 (2015-12-10)
==========================

- Added check for driver present to util app
- Workaround for multiple null module load
- Added defines for common stream formats
- Support dual build of python2 and python3
- Improved module loader API and results query
- New query information for SoapySDRUtil --probe
- Added setLogLevel() and default log threshold
- Added API to query native stream format
- Added C++ typedef for list of Kwargs
- Added API to query tune argument info
- Added API to query sensors meta info
- Added API to query setting argument info
- Added API to query stream argument info
- CMAKE_BUILD_TYPE automatically defaults to Release
- Added API to query AGC mode availability
- Added API to query available clock rates
- Set library SOVERSION using SOAPY_SDR_ABI_VERSION

Release 0.3.1 (2015-10-11)
==========================

- Fixed missing python bindings for Errors.hpp
- Added -DSOAPY_SDR_ROOT=<path> option to build

Release 0.3.0 (2015-10-10)
==========================

- Util --probe option for detailed info summary
- Added corrections API to check for availability
- Added API to convert error codes to string
- Added per-channel API to query readback sensors
- Added GPIO API data direction modification mask
- Removed SoapyUHD and SoapyOsmo submodules

Release 0.2.3 (2015-10-08)
==========================

- Append to linker flags to avoid overwriting them
- Link with -flat_namespace for OSX application bundle
- GCC link with -pthread for threaded client support

Release 0.2.2 (2015-08-15)
==========================

- Fixed vector<double> in python swig bindings
- Support CMake 2.8.7 builds (ubuntu 12.04 lts)

Release 0.2.1 (2015-07-09)
==========================

- Fixed Device::make() factory argument forwarding
- Added Time.h/Time.hpp time conversion API calls
- Implement SoapySDRConfigVersion.cmake version check

Release 0.2.0 (2015-06-15)
==========================

- Support /usr/local module installs when root is /usr
- Support SOAPY_SDR_PLUGIN_PATH environment variable
- Added logger support for stream status indicator
- Frequency component API and tune distribution algorithm
- Added direct buffer access API for read and write streams
- Automatic gain distribution and overall range calculations
- Added read stream status Device API call
- Added additional error codes and status flags

Release 0.1.1 (2015-06-14)
==========================

- Fix path issues for listModules() implementation under MSVC
- Support openSuSE for automatic LIB_SUFFIX detection logic
- patch for missing add_compile_options in older CMake

Release 0.1.0 (2015-01-31)
==========================

This is the first public release of the SoapySDR project.


apps/SoapySDRUtil.cpp
// Copyright (c) 2014-2016 Josh Blum
// SPDX-License-Identifier: BSL-1.0

#include <SoapySDR/Version.hpp>
#include <SoapySDR/Modules.hpp>
#include <SoapySDR/Registry.hpp>
#include <SoapySDR/Device.hpp>
#include <cstdlib>
#include <cstddef>
#include <iostream>
#include <getopt.h>

std::string SoapySDRDeviceProbe(SoapySDR::Device *);

/***********************************************************************
 * Print help message
 **********************************************************************/
static int printHelp(void)
{
    std::cout << "Usage SoapySDRUtil [options]" << std::endl;
    std::cout << "  Options summary:" << std::endl;
    std::cout << "    --help \t\t\t\t Print this help message" << std::endl;
    std::cout << "    --info \t\t\t\t Print module information" << std::endl;
    std::cout << "    --find[=\"driver=foo,type=bar\"] \t Discover available devices" << std::endl;
    std::cout << "    --make[=\"driver=foo,type=bar\"] \t Create a device instance" << std::endl;
    std::cout << "    --probe[=\"driver=foo,type=bar\"] \t Print detailed information" << std::endl;
    std::cout << "    --check[=driverName] \t\t Check if driver is present" << std::endl;
    std::cout << std::endl;

    std::cout << "  Rate testing options:" << std::endl;
    std::cout << "    --args[=\"driver=foo\"] \t\t Arguments for testing" << std::endl;
    std::cout << "    --rateTest[=rate Sps] \t\t Rate in samples per second" << std::endl;
    std::cout << "    --channels[=\"0, 1, 2\"] \t\t List of channels, default 0" << std::endl;
    std::cout << "    --direction[=RX or TX] \t\t Specify the channel direction" << std::endl;
    std::cout << std::endl;
    return EXIT_SUCCESS;
}

/***********************************************************************
 * Print version and module info
 **********************************************************************/
static int printInfo(void)
{
    std::cout << "Lib Version: v" << SoapySDR::getLibVersion() << std::endl;
    std::cout << "API Version: v" << SoapySDR::getAPIVersion() << std::endl;
    std::cout << "ABI Version: v" << SoapySDR::getABIVersion() << std::endl;
    std::cout << "Install root: " << SoapySDR::getRootPath() << std::endl;

    for (const auto &path : SoapySDR::listSearchPaths())
        std::cout << "Search path: " << path << std::endl;

    const auto modules = SoapySDR::listModules();
    for (const auto &mod : modules) std::cout << "Module found: " << mod << std::endl;
    if (modules.empty()) std::cout << "No modules found!" << std::endl;

    std::cout << "Loading modules... " << std::flush;
    SoapySDR::loadModules();
    std::cout << "done" << std::endl;

    std::cout << "Available factories...";
    const auto factories = SoapySDR::Registry::listFindFunctions();
    for (const auto &it : factories) std::cout << it.first << ", ";
    if (factories.empty()) std::cout << "No factories found!" << std::endl;
    std::cout << std::endl;
    return EXIT_SUCCESS;
}

/***********************************************************************
 * Find devices and print args
 **********************************************************************/
static int findDevices(void)
{
    std::string argStr;
    if (optarg != NULL) argStr = optarg;

    const auto results = SoapySDR::Device::enumerate(argStr);
    for (size_t i = 0; i < results.size(); i++)
    {
        std::cout << "Found device " << i << std::endl;
        for (const auto &it : results[i])
        {
            std::cout << "  " << it.first << " = " << it.second << std::endl;
        }
        std::cout << std::endl;
    }
    if (results.empty())
    {
        std::cerr << "No devices found!" << std::endl;
        return EXIT_FAILURE;
    }
    std::cout << std::endl;
    return EXIT_SUCCESS;
}

/***********************************************************************
 * Make device and print hardware info
 **********************************************************************/
static int makeDevice(void)
{
    std::string argStr;
    if (optarg != NULL) argStr = optarg;

    std::cout << "Make device " << argStr << std::endl;
    try
    {
        auto device = SoapySDR::Device::make(argStr);
        std::cout << "  driver=" << device->getDriverKey() << std::endl;
        std::cout << "  hardware=" << device->getHardwareKey() << std::endl;
        for (const auto &it : device->getHardwareInfo())
        {
            std::cout << "  " << it.first << "=" << it.second << std::endl;
        }
        SoapySDR::Device::unmake(device);
    }
    catch (const std::exception &ex)
    {
        std::cerr << "Error making device: " << ex.what() << std::endl;
        return EXIT_FAILURE;
    }
    std::cout << std::endl;
    return EXIT_SUCCESS;
}

/***********************************************************************
 * Make device and print detailed info
 **********************************************************************/
static int probeDevice(void)
{
    std::string argStr;
    if (optarg != NULL) argStr = optarg;

    std::cout << "Probe device " << argStr << std::endl;
    try
    {
        auto device = SoapySDR::Device::make(argStr);
        std::cout << SoapySDRDeviceProbe(device) << std::endl;
        SoapySDR::Device::unmake(device);
    }
    catch (const std::exception &ex)
    {
        std::cerr << "Error probing device: " << ex.what() << std::endl;
        return EXIT_FAILURE;
    }
    std::cout << std::endl;
    return EXIT_SUCCESS;
}

/***********************************************************************
 * Check the registry for a specific driver
 **********************************************************************/
static int checkDriver(void)
{
    std::string driverName;
    if (optarg != NULL) driverName = optarg;

    std::cout << "Loading modules... " << std::flush;
    SoapySDR::loadModules();
    std::cout << "done" << std::endl;

    std::cout << "Checking driver '" << driverName << "'... " << std::flush;
    const auto factories = SoapySDR::Registry::listFindFunctions();

    if (factories.find(driverName) == factories.end())
    {
        std::cout << "MISSING!" << std::endl;
        return EXIT_FAILURE;
    }
    else
    {
        std::cout << "PRESENT" << std::endl;
        return EXIT_SUCCESS;
    }
}

/***********************************************************************
 * main utility entry point
 **********************************************************************/
int main(int argc, char *argv[])
{
    std::cout << "######################################################" << std::endl;
    std::cout << "## Soapy SDR -- the SDR abstraction library" << std::endl;
    std::cout << "######################################################" << std::endl;
    std::cout << std::endl;

    /*******************************************************************
     * parse command line options
     ******************************************************************/
    static struct option long_options[] = {
        {"help", no_argument, 0, 'h'},
        {"find", optional_argument, 0, 'f'},
        {"make", optional_argument, 0, 'm'},
        {"info", optional_argument, 0, 'i'},
        {"probe", optional_argument, 0, 'p'},
        {"check", optional_argument, 0, 'c'},
        {0, 0, 0,  0}
    };
    int long_index = 0;
    int option = 0;
    while ((option = getopt_long_only(argc, argv, "", long_options, &long_index)) != -1)
    {
        switch (option)
        {
        case 'h': return printHelp();
        case 'i': return printInfo();
        case 'f': return findDevices();
        case 'm': return makeDevice();
        case 'p': return probeDevice();
        case 'c': return checkDriver();
        }
    }

    //unknown or unspecified options, do help...
    return printHelp();
}
